File: automatepro_cpp_tutorials/src/io/digital_drive_out.cpp
#include <chrono>
#include <memory>
#include <mutex>

#include <rclcpp/rclcpp.hpp>
#include <automatepro_interfaces/msg/digital_drive_out.hpp>

class DigitalDriveOutPublisher : public rclcpp::Node
{
public:
  DigitalDriveOutPublisher()
  : Node("digital_drive_out_publisher"), duty_cycle_(0)
  {
    publisher_ = this->create_publisher<automatepro_interfaces::msg::DigitalDriveOut>(
      "/io/digital_drive_out", 10);
    timer_ = this->create_wall_timer(
      std::chrono::seconds(1),
      std::bind(&DigitalDriveOutPublisher::timer_callback, this));
  }

  void switch_off()
  {
    std::lock_guard<std::mutex> lock(mutex_);
    stopped_ = true;
    timer_->cancel();
    auto msg = automatepro_interfaces::msg::DigitalDriveOut();
    msg.d_drive_pin_id = automatepro_interfaces::msg::DigitalDriveOut::HALF_BRIDGE_DRIVE_01;
    msg.direction = automatepro_interfaces::msg::DigitalDriveOut::FORWARD;
    msg.duty_cycle_percent = 0;
    publisher_->publish(msg);
    // Fast DDS acknowledges on the writer heartbeat, sent every 3 s by default.
    if (publisher_->wait_for_all_acked(std::chrono::seconds(4))) {
      RCLCPP_INFO(this->get_logger(), "Switched HALF_BRIDGE_DRIVE_01 off");
    } else {
      RCLCPP_WARN(this->get_logger(), "HALF_BRIDGE_DRIVE_01 off command not acknowledged");
    }
  }

private:
  void timer_callback()
  {
    std::lock_guard<std::mutex> lock(mutex_);
    if (stopped_) {
      return;
    }
    auto msg = automatepro_interfaces::msg::DigitalDriveOut();
    msg.d_drive_pin_id = automatepro_interfaces::msg::DigitalDriveOut::HALF_BRIDGE_DRIVE_01;
    msg.direction = automatepro_interfaces::msg::DigitalDriveOut::FORWARD;
    msg.duty_cycle_percent = duty_cycle_;
    publisher_->publish(msg);
    RCLCPP_INFO(
      this->get_logger(), "Publishing DigitalDriveOut: d_out_pin_id=%d, direction=%d, duty_cycle_percent=%d",
      msg.d_drive_pin_id, msg.direction, msg.duty_cycle_percent);
    duty_cycle_ = (duty_cycle_ == 0) ? 100 : 0;     // Toggle duty cycle between 0(ON) and 100(OFF)
  }

  rclcpp::Publisher<automatepro_interfaces::msg::DigitalDriveOut>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  int duty_cycle_;
  std::mutex mutex_;
  bool stopped_{false};
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<DigitalDriveOutPublisher>();
  // The IO controller keeps the last command, so switch the output off before shutdown.
  std::weak_ptr<DigitalDriveOutPublisher> weak_node = node;
  rclcpp::contexts::get_global_default_context()->add_pre_shutdown_callback(
    [weak_node]() {
      if (auto locked_node = weak_node.lock()) {
        locked_node->switch_off();
      }
    });
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: automatepro_python_tutorials/automatepro_python_tutorials/io/digital_drive_out.py
import signal

import rclpy
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.signals import SignalHandlerOptions
from automatepro_interfaces.msg import DigitalDriveOut


class DigitalDriveOutPublisher(Node):

    def __init__(self):
        super().__init__('digital_drive_out_publisher')
        self.publisher_ = self.create_publisher(DigitalDriveOut, '/io/digital_drive_out', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)  # 1s
        self.duty_cycle = 0

    def timer_callback(self):
        msg = DigitalDriveOut()
        msg.d_drive_pin_id = DigitalDriveOut.HALF_BRIDGE_DRIVE_01
        msg.direction = DigitalDriveOut.FORWARD
        msg.duty_cycle_percent = self.duty_cycle
        self.publisher_.publish(msg)
        self.get_logger().info(
            'Publishing DigitalDriveOut: d_drive_pin_id=%d, direction=%d, '
            'duty_cycle_percent=%d' %
            (msg.d_drive_pin_id, msg.direction, msg.duty_cycle_percent))
        # Toggle duty cycle between 0 (OFF) and 100 (ON)
        self.duty_cycle = 100 if self.duty_cycle == 0 else 0

    def switch_off(self):
        msg = DigitalDriveOut()
        msg.d_drive_pin_id = DigitalDriveOut.HALF_BRIDGE_DRIVE_01
        msg.direction = DigitalDriveOut.FORWARD
        msg.duty_cycle_percent = 0
        self.publisher_.publish(msg)
        # With no matched subscriber, wait_for_all_acked() succeeds without delivering anything.
        if self.publisher_.get_subscription_count() == 0:
            self.get_logger().warn(
                'HALF_BRIDGE_DRIVE_01 off command not delivered: no subscriber on %s'
                % self.publisher_.topic_name)
            return
        # Fast DDS acknowledges on the writer heartbeat, sent every 3 s by default.
        if self.publisher_.wait_for_all_acked(Duration(seconds=4)):
            self.get_logger().info('Switched HALF_BRIDGE_DRIVE_01 off')
        else:
            self.get_logger().warn('HALF_BRIDGE_DRIVE_01 off command not acknowledged')


def main(args=None):
    # The IO controller keeps the last command, so switch the output off before exiting.
    rclpy.init(args=args, signal_handler_options=SignalHandlerOptions.NO)
    signal.signal(signal.SIGTERM, signal.default_int_handler)
    node = DigitalDriveOutPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.switch_off()
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
